### PR TITLE
feat: notify admin final CMO draft ready for judge (FPLA-927)

### PR DESCRIPTION
**Pull request by Jamesnelson117**
**Before creating a pull request make sure that:**

- [x] PR's title includes JIRA ticket number and follow [Conventional Commits](https://www.conventionalcommits.org/) specification
- [x] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-927

### Change description ###
Notification template added to send a notification when an LA readies a CMO for Judge

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java [hunk 1]
```diff
@@ -305,13 +310,41 @@ void shouldNotifyRepresentativesOfCMOIssued() throws Exception {
                 eq(expectedCMOIssuedNotificationParametersForRepresentative), eq("12345"));
         }
 
+        @Test
+        void shouldNotifyAdminOfCMOReadyForJudgeReview() throws Exception {
+            CallbackRequest callbackRequest = callbackRequest();
+            CaseDetails caseDetails = callbackRequest().getCaseDetails();
+
+            given(hmctsCourtLookupConfiguration.getCourt(LOCAL_AUTHORITY_CODE))
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java]: Do we need to do this given and the one below as well, I thought that we created some configuration beans in the so we didn't need to do this

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java [hunk 2]
```diff
@@ -68,31 +96,88 @@ void aboutToSubmitShouldPopulateListServedCaseManagementOrdersWhenSendsToAllPart
...
+                .putAll(caseDataMap(order))
+                .putAll(getNotificationData())
+                .build())
+            .id(CASE_ID)
+            .build();
+
+        postAboutToSubmitEvent(buildCallbackRequest(caseDetails, DRAFT_CASE_MANAGEMENT_ORDER));
+
+        verify(notificationClient, times(1)).sendEmail(
+            eq(CMO_READY_FOR_JUDGE_REVIEW_NOTIFICATION_TEMPLATE), eq("admin@family-court.com"),
```
Comment (Gogetter) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java]: `times(1)` not required as by default `verify(...)` does calls `times(1)`

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java [hunk 3]
```diff
@@ -68,31 +96,88 @@ void aboutToSubmitShouldPopulateListServedCaseManagementOrdersWhenSendsToAllPart
...
+        AboutToStartOrSubmitCallbackResponse response = postAboutToSubmitEvent(buildCallbackRequest(caseDetails,
+            ACTION_CASE_MANAGEMENT_ORDER));
 
         assertThat(response.getData())
             .containsOnlyKeys(SERVED_CASE_MANAGEMENT_ORDERS.getKey(), HEARING_DETAILS_KEY);
     }
 
-    private Map<String, Object> caseDataMap(CaseManagementOrder order, LocalDateTime localDateTime) {
+    @Test
+    void aboutToSubmitShouldBuildNotificationTemplateWhenSentToJudge() throws Exception {
```
Comment (Gogetter) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java]: looks like we are also sending emails?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java [hunk 4]
```diff
@@ -68,31 +96,88 @@ void aboutToSubmitShouldPopulateListServedCaseManagementOrdersWhenSendsToAllPart
...
                 .build()));
     }
 
-    private CallbackRequest buildCallbackRequest(CaseDetails caseDetails) {
+    private Map<String, Object> getNotificationData() {
+        return ImmutableMap.of(
+            "cmoEventId", "draftCMO",
+            "caseLocalAuthority", LOCAL_AUTHORITY_CODE,
+            "familyManCaseNumber", FAMILY_MAN_CASE_NUMBER,
+            "respondents1", ImmutableList.of(
```
Comment (Gogetter) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java]: why not `respondentLastName` going by the method name?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java [hunk 5]
```diff
@@ -68,31 +96,88 @@ void aboutToSubmitShouldPopulateListServedCaseManagementOrdersWhenSendsToAllPart
...
+            .build();
+    }
+
+    private String buildSubjectLine() {
+        String hearingDate = dateFormatterService
+            .formatLocalDateTimeBaseUsingFormat(YESTERDAY, "d MMM yyyy");
+
+        return String.format("%s, %s, hearing %s", RESPONDENT_SURNAME, FAMILY_MAN_CASE_NUMBER, hearingDate);
+    }
 }
```
Comment (Gogetter) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java]: would we need a test scenario for when we do not have a value for `hearingDate`? Or would we always have this value?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderProgressionServiceTest.java [hunk 6]
```diff
@@ -50,11 +54,20 @@
...
+
+    @Mock
+    private RequestData requestData;
+
     private CaseManagementOrderProgressionService service;
 
     @BeforeEach
     void setUp() {
-        this.service = new CaseManagementOrderProgressionService(mapper);
+        given(requestData.authorisation()).willReturn("Bearer");
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderProgressionServiceTest.java]: Are these 2 lines needed?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderProgressionService.java [hunk 7]
```diff
@@ -27,10 +31,16 @@
...
 
     private final ObjectMapper mapper;
+    private final RequestData requestData;
+    private final ApplicationEventPublisher applicationEventPublisher;
 
     @Autowired
-    public CaseManagementOrderProgressionService(ObjectMapper mapper) {
+    public CaseManagementOrderProgressionService(ObjectMapper mapper,
+                                                RequestData requestData,
+                                                ApplicationEventPublisher applicationEventPublisher) {
```
Comment (Gogetter) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderProgressionService.java]: I think we might want to consider `@RequiredArgsConstructor(onConstructor = @__(@Autowired))` in place of writing out the `@Autowired` constructor

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderProgressionService.java [hunk 8]
```diff
@@ -48,6 +58,7 @@ private void progressDraftCaseManagementOrder(CaseDetails caseDetails, CaseManag
             case SEND_TO_JUDGE:
                 caseDetails.getData().put(CASE_MANAGEMENT_ORDER_JUDICIARY.getKey(), order);
                 caseDetails.getData().remove(CASE_MANAGEMENT_ORDER_LOCAL_AUTHORITY.getKey());
+                publishReadyForJudgeReviewEvent(caseDetails, requestData);
```
Comment (Gogetter) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderProgressionService.java]: just curious, why not use `submitted` in  `ActionCaseManagementOrderController`?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java [hunk 9]
```diff
@@ -305,13 +310,41 @@ void shouldNotifyRepresentativesOfCMOIssued() throws Exception {
...
 
+        private ImmutableMap<String, Object> getCMOReadyForJudgeNotificationParameters() {
+            return ImmutableMap.<String, Object>builder()
+                .putAll(expectedCommonCMONotificationParameters())
+                .build();
+        }
+
         private CaseDetails buildCaseDetailsWithRepresentatives() throws IOException {
             CaseDetails caseDetails = callbackRequest().getCaseDetails();
             Map<String, Object> caseData = caseDetails.getData();
```
Comment (Gogetter) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java]: do we need a `shouldNot` scenario?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderProgressionServiceTest.java (hunk 6 above)
Comment (lewisbirks) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderProgressionServiceTest.java]: Also I know you haven't touched the method but `caseDataWithCaseManagementOrder` doesn't need to throw an IOException

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java [hunk 10]
```diff
@@ -68,31 +96,88 @@ void aboutToSubmitShouldPopulateListServedCaseManagementOrdersWhenSendsToAllPart
...
+    private Map<String, Object> expectedTemplateParameters() {
+        return ImmutableMap.<String, Object>builder()
+            .put("subjectLineWithHearingDate", buildSubjectLine())
+            .put("reference", CASE_REFERENCE)
+            .put("respondentLastName", RESPONDENT_SURNAME)
+            .put("caseUrl", "http://fake-url/case/" + JURISDICTION + "/" + CASE_TYPE + "/12345")
+            .build();
+    }
+
+    private String buildSubjectLine() {
```
Comment (Gogetter) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java]: why not reuse `EmailNotificationHelper.buildSubjectLine`?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java (hunk 4 above)
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java]: `respondents1` is the property we store it on in CCD caseData

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java (hunk 10 above)
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java]: Discussed

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java (hunk 5 above)
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseManagementOrderProgressionControllerTest.java]: No, we have validation to ensure the user cannot get to this point without having at least one hearingDate

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderProgressionService.java (hunk 8 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderProgressionService.java]: One for @swalker125

Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderProgressionService.java]: I believe the reason was that as this handles the final flow of the CMO object (where it goes i.e. final -> draft, draft -> final...) it was decided to do it here instead, not 100% sure on that though

**Pull request merged**